### DISCUSSION TITLE
Add validation accuracy during training

### DISCUSSION THREAD

**Discussion by fgraffitti-cyberhawk** [facebookresearch/detectron2#4557] [Q&A]
I'm trying to add some information on the validation dataset during training.
While adding losses is quite straightforward (running the model in training mode on the items of the validation data loader returns a dictionary with the losses), I am finding quite tricky to calculate the validation accuracy.

For maskRCNN for example, it seems to me that the training accuracy is computed in [BaseMaskRCNNHead.forward](https://github.com/facebookresearch/detectron2/blob/32b61e64c76118b2e9fc2237f283a8e9c938bd16/detectron2/modeling/roi_heads/mask_head.py#L183) using the [mask_rcnn_loss ](https://github.com/facebookresearch/detectron2/blob/32b61e64c76118b2e9fc2237f283a8e9c938bd16/detectron2/modeling/roi_heads/mask_head.py#L33) function, but I can't find a way to use this (or a similar) approach for calculating the validation accuracy.

Anyone has an idea on how to approach this problem?

**Comment by aymanaboghonim** [comment]
you can add a val hook like this code 
`class ValidationLoss(HookBase):
    def __init__(self, cfg):
        super().__init__()
        self.cfg = cfg.clone()
        self.cfg.DATASETS.TRAIN = cfg.DATASETS.TEST
        self._loader = iter(build_detection_train_loader(self.cfg))
        
    def after_step(self):
        data = next(self._loader)
        with torch.no_grad():
            loss_dict = self.trainer.model(data)
            
            losses = sum(loss_dict.values())
            assert torch.isfinite(losses).all(), loss_dict

            loss_dict_reduced = {"val_" + k: v.item() for k, v in 
                                 comm.reduce_dict(loss_dict).items()}
            losses_reduced = sum(loss for loss in loss_dict_reduced.values())
            if comm.is_main_process():
                self.trainer.storage.put_scalars(total_val_loss=losses_reduced, 
                                                **loss_dict_reduced)
val_loss = ValidationLoss(cfg) `

then register the hook like this 
`trainer.register_hooks([val_loss])
trainer._hooks = trainer._hooks[:-2] + trainer._hooks[-2:][::-1]
trainer.resume_or_load(resume=False)
trainer.train()` to start training and get val metrics

**Reply by fgraffitti-cyberhawk** [reply]
hey @aymanaboghonim , thanks for your reply, I'll give it a try later. I have a question regarding your code though.
It looks like your code allows you to calculate the validation losses but not the validation accuracy, is thins correct?
I already have implemented validation losses in my code, and now I was looking at implementing the accuracy on the val set as well.

**Reply by aymanaboghonim** [reply]
sorry for replaying late. yes, it is about loss not accuracy.
have you coded the val accuracy or not yet ??

**Reply by fgraffitti-cyberhawk** [reply]
no, I haven't written any code for accuracy yet, got busy working on other stuff :/ I'm planning to do that at some point in the future but haven't had a chance yet :)
